secret-rotator/tests/mockclient.go

```diff
@@ -0,0 +1,255 @@
+	_, ok := cl.Secrets[project][id]
+	if !ok {
+		return status.Error(codes.NotFound, fmt.Sprintf("Secret [projects/%s/secrets/%s] not found or has no versions.", project, id))
+	}
+
+	return nil
+}
+
+// ValidateSecretVersion returns nil if the secret version exists, otherwise error.
+func (cl *MockClient) ValidateSecretVersion(project, id, version string) error {
+	err := cl.ValidateSecret(project, id)
+	if err != nil {
+		return err
+	}
+
+	if version == "latest" {
+		version = strconv.Itoa(len(cl.Secrets[project][id].Versions))
+	}
+
+	_, ok := cl.Secrets[project][id].Versions[version]
+	if !ok {
+		return status.Error(codes.NotFound, fmt.Sprintf("Secret Version [projects/%s/secrets/%s/versions/%s] not found.", project, id, version))
+	}
+
+	return nil
+}
+
+// UpsertSecret adds a new version to the secret specified by project, id.
+// It inserts a new secret if id doesn't already exist.
+// If successful the latest version will have 'data' as its secret value,
+// and return the latest version number if successful, otherwise returns error
+func (cl *MockClient) UpsertSecret(project, id string, data []byte) (string, error) {
+	err := cl.ValidateProject(project)
+	if err != nil {
+		return "", err
+	}
+
+	err = cl.ValidateSecret(project, id)
+	if err != nil {
+		cl.Secrets[project][id] = new(Secret)
+	}
+
+	version := strconv.Itoa(len(cl.Secrets[project][id].Versions) + 1)
+	cl.Secrets[project][id].Versions[version] = &Version{
+		Data:  data,
+		State: secretmanagerpb.SecretVersion_ENABLED,
+	}
+
+	return version, nil
+}
+
+// GetCreateTime gets the createTime of the secret version specified by project, id, version.
+// Returns createTime if successful, otherwise error.
+func (cl *MockClient) GetCreateTime(project, id, version string) (time.Time, error) {
+	err := cl.ValidateSecretVersion(project, id, version)
+	if err != nil {
+		return time.Time{}, err
+	}
+
+	if version == "latest" {
+		version = strconv.Itoa(len(cl.Secrets[project][id].Versions))
+	}
+
+	creatTime := cl.Secrets[project][id].Versions[version].CreateTime
+
+	return creatTime, nil
+}
+
+// GetSecretLabels gets the labels of the secret specified by project, id.
+// Returns secret labels if successful, otherwise error
+func (cl *MockClient) GetSecretLabels(project, id string) (map[string]string, error) {
+	err := cl.ValidateSecret(project, id)
+	if err != nil {
+		return nil, err
+	}
+
+	ret := make(map[string]string)
+	for key, val := range cl.Secrets[project][id].Labels {
+		ret[key] = val
+	}
+
+	return ret, nil
+}
+
+// GetSecretVersionData gets the data of the secret version specified by project, id, version.
+// Returns secret value if successful, otherwise error
+func (cl *MockClient) GetSecretVersionData(project, id, version string) ([]byte, error) {
+	err := cl.ValidateSecretVersion(project, id, version)
+	if err != nil {
+		return nil, err
+	}
+
+	if version == "latest" {
+		version = strconv.Itoa(len(cl.Secrets[project][id].Versions))
+	}
```

Comment:
For followup: this pattern seems reused a lot in this file.  Is there some way you could reuse this and avoid the copy-paste?

Reply:
Added an utility function `ValidateAndConvertVersion()` that checks if the secret version exists and also converts the version into the string of version number if version is "latest".